### PR TITLE
#101 Migration improvements logging

### PR DESCRIPTION
**Pull request by adrianchivu**
### Prerequisites

- [X] I have written a descriptive pull-request title
- [X] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/CDP4-Server-Administration-Tool/pulls) open
- [X] I have verified that I am following the CDP4-Server-Administration-Tool [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/CDP4-Server-Administration-Tool/master/.github/CONTRIBUTING.md)
- [X] I have provided test coverage for my change (where applicable)

### CODE REVIEW

File: Modules/Migration/ViewModels/MigrationViewModel.cs [hunk 1]
```diff
@@ -164,6 +172,28 @@ public void AddSubscriptions()
             {
                 this.MigrationFactory.TargetSession = this.TargetViewModel.ServerSession;
             });
+
+            CDPMessageBus.Current.Listen<LogEvent>().Subscribe((operationEvent) => {
```
Comment (cozminvelciu) [inline Modules/Migration/ViewModels/MigrationViewModel.cs:176]: Is there any way to not duplicate the logging logic both here and in `Migration.cs`?

File: Modules/Migration/ViewModels/MigrationViewModel.cs [hunk 2]
```diff
@@ -319,6 +349,8 @@ private void OperationMessageHandler(string message)
             if (string.IsNullOrEmpty(message)) return;
 
             this.Output += $"{DateTime.Now:HH:mm:ss} {message}{Environment.NewLine}";
+
+            Logger.Debug(message);
```
Comment (cozminvelciu) [inline Modules/Migration/ViewModels/MigrationViewModel.cs:353]: Will this log the Migration log messages twice, as they are also logged there?

File: Modules/Common/ViewModels/LoginViewModel.cs [hunk 3]
```diff
@@ -413,11 +415,22 @@ private void ExecuteLoadSourceFile()
...
         /// </summary>
-        /// <param name="message"></param>
-        private void LogMessage(string message)
+        /// <param name="message">Message that will be logged</param>
+        /// <param name="ex">Exception that will be logged</param>
+        private void LogMessage(string message, Exception ex = null)
         {
-            Debug.WriteLine(message);
-            this.Output = message;
+            var logMessage = ex?.Message != null ? message + ex.Message : message;
```
Comment (cozminvelciu) [inline Modules/Common/ViewModels/LoginViewModel.cs:422]: Should this be formatted the same way as the other error messages?

File: Modules/Migration/Views/Layout.xaml.cs [hunk 4]
```diff
@@ -37,5 +41,21 @@ public Layout()
...
+        /// <param name="e">The <see cref="EditValueChangedEventArgs" /></param>
+        private void BaseEdit_OnEditValueChanged(object sender, EditValueChangedEventArgs e)
+        {
+            if (!(sender is TextEdit textEdit))
+            {
+                return;
+            }
+
+            textEdit.Focus();
+            Dispatcher.BeginInvoke(new Action(() => textEdit.SelectionStart = textEdit.Text.Length));
```
[the remark this answers is not in the document]
Comment (adrianchivu) [reply Modules/Migration/Views/Layout.xaml.cs:58]: For the moment I don't have a solution.

File: Modules/Migration/Utils/Migration.cs [hunk 5]
```diff
@@ -214,19 +141,30 @@ public async Task<bool> ImportData(List<EngineeringModelRowViewModel> selectedMo
...
-                this.NotifyMessage("Please select model(s) to migrate.");
+                CDPMessageBus.Current.SendMessage(new LogEvent
+                {
+                    Message = "Please select model(s) to migrate."
+                });
+
                 return false;
             }
 
             this.NotifyStep(MigrationStep.ImportStart);
```
Comment (cozminvelciu) [inline Modules/Migration/Utils/Migration.cs:164]: Do we still need the `NotifyStep` logic? Could we also use the `CDPMessageBus` for this?

File: Modules/Common/ViewModels/LoginViewModel.cs [hunk 6]
```diff
@@ -412,26 +424,7 @@ private void ExecuteLoadSourceFile()
...
-                Logger.Info(logMessage);
-            }
-            else
-            {
-                Logger.Error(logMessage);
-            }
-
-            this.Output = logMessage;
-        }
+
```
Comment (cozminvelciu) [inline Modules/Common/ViewModels/LoginViewModel.cs:427]: Two unneeded extra newlines.

File: Modules/Migration/Utils/Migration.cs [hunk 7]
```diff
@@ -239,8 +233,10 @@ public async Task<bool> ImportData(List<EngineeringModelRowViewModel> selectedMo
                 }
             }
 
-            this.NotifyStep(MigrationStep.ImportEnd);
-
+            CDPMessageBus.Current.SendMessage(new LogEvent
+            {
+                Message = "Import operation end"
+            });
             return true;
```
Comment (cozminvelciu) [inline Modules/Migration/Utils/Migration.cs:240]: Missing newline.

**Review by cozminvelciu — APPROVED** [review]

**Pull request merged**